File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    vector<string> key = {
        "", "", "abc", "def",
        "ghi", "jkl", "mno",
        "pqrs", "tuv", "wxyz"
    };

    void solve(string digits, vector<string> &ans, string current) {
        if (digits.length() == 0) {
            ans.push_back(current);
            return;
        }

        string letters = key[digits[0] - '0'];

        for (int i = 0; i < letters.length(); i++) {
            solve(digits.substr(1), ans, current + letters[i]);
        }
    }

    vector<string> letterCombinations(string digits) {
        vector<string> ans;

        if (digits.length() == 0)
            return ans;

        solve(digits, ans, "");
        return ans;
    }
};

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {
    static String[] key = {"", "", "abc", "def","ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
    void solve(String d, List<String> ans,String s){
         if(d.length()==0){
            ans.add(s);
            return;
         }

         String ques = key[(int)(d.charAt(0))-48];
         for(int i=0; i<ques.length(); i++){
            solve(d.substring(1), ans, s+ques.charAt(i));
         }
    }
    public List<String> letterCombinations(String digits) {
        
        List<String> ans = new ArrayList<String>();
        if(digits.length()==0){
            return ans;
        }
        solve(digits, ans, "");
        return ans;

    }

}